hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FairScheduler.java

```diff
@@ -987,9 +978,19 @@ void attemptScheduling(FSSchedulerNode node) {
       }
 
       // Assign new containers...
-      // 1. Check for reserved applications
-      // 2. Schedule if there are no reservations
-
+      // 1. Ensure containers are assigned to the apps that preempted
+      // 2. Check for reserved applications
+      // 3. Schedule if there are no reservations
+
+      // Apps may wait for preempted containers
+      // We have to satisfy these first to avoid cases, when we preempt
+      // a container for A from B and C gets the preempted containers,
+      // when C does not qualify for preemption itself.
+      for (FSAppAttempt app : node.getPreemptionList()) {
```

Comment:
We don't seem to handle the case where an app is to be allocated multiple containers from based on the preempted resources.

Would it help to store the apps in a TreeMap in FSSchedulerNode and have a method that returns the next preempted app to allocate to? If none exist, return null? Also, if this were to work, we might have to revert back to the behavior in an earlier patch: add app/resources to the map only after the container is actually preempted?